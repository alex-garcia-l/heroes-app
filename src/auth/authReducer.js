import { authType } from '../types/auth';

export const authReducer = (state = {}, action) => {
  
  switch (action.type) {
    case authType.login:
      return {
        ...action.payload,
        logged: true
      }
    case authType.logout:
      return {
        logged: false
      }
  
    default:
      return state;
  }

}