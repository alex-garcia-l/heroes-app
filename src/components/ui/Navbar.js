import React from 'react';
import { Link, NavLink, useNavigate } from 'react-router-dom';

export const Navbar = () => {

  const navigate = useNavigate();

  const handleClick = (evt) => {
    evt.preventDefault();
    navigate('/login', {
      replace: true
    });
  }
  return (
    <header className="p-3 bg-dark text-white">
      <div className="container">
        <div className="d-flex flex-wrap align-items-center justify-content-center justify-content-lg-start">
          <Link to="/" className="d-flex align-items-center mb-2 mb-lg-0 text-white text-decoration-none me-0 me-md-4">
            <span className="fs-4">Heroes App</span>
          </Link>

          <ul className="nav col-12 col-lg-auto me-lg-auto mb-2 justify-content-center mb-md-0">
            <li>
              <NavLink
                to="/"
                className={({ isActive }) => `nav-link px-2 ${(isActive) ? 'text-secondary' : 'text-white'}`}
              >
                Home
              </NavLink>
            </li>
            <li>
              <NavLink
                to="/marvel"
                className={({ isActive }) => `nav-link px-2 ${(isActive) ? 'text-secondary' : 'text-white'}`}
              >
                Marvel
              </NavLink>
            </li>
            <li>
              <NavLink
                to="/dc"
                className={({ isActive }) => `nav-link px-2 ${(isActive) ? 'text-secondary' : 'text-white'}`}
              >
                Dc
              </NavLink>
            </li>
          </ul>

          <div className="text-end">
            <span className="me-4">Name</span>
            <button
              type="button"
              className="btn btn-outline-light me-2"
              onClick={handleClick}
            >
              Logout
            </button>
          </div>
        </div>
      </div>
    </header>
  );
};
