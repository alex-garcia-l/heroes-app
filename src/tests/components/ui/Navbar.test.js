import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../../auth/authContext';
import { Navbar } from '../../../components/ui/Navbar';
import { authType } from '../../../types/auth';


const userDispatch = jest.fn();
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));

describe('Pruebas en el componente <Navbar />', () => {

  const contextValue = {
    user: {
      name: 'Name',
      logged: true
    },
    userDispatch
  }

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={['/home']}>
        <Navbar />
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test('debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span.me-4').text().trim()).toBe(contextValue.user.name);
  });

  test('debe llamar el logout, llamar el navigate y el userDispatch con los argumentos', () => {
    wrapper.find('button').simulate('click', {
      preventDefault() { }
    });

    expect(userDispatch).toHaveBeenCalledWith({ type: authType.logout });
    expect(mockNavigate).toHaveBeenCalledWith('/login', {
      replace: true
    });
  });

});
