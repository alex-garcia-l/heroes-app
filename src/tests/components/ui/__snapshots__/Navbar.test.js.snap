// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Navbar /> debe mostrarse correctamente 1`] = `
<header
  className="p-3 bg-dark text-white"
>
  <div
    className="container"
  >
    <div
      className="d-flex flex-wrap align-items-center justify-content-center justify-content-lg-start"
    >
      <a
        className="d-flex align-items-center mb-2 mb-lg-0 text-white text-decoration-none me-0 me-md-4"
        href="/home"
        onClick={[Function]}
      >
        <span
          className="fs-4"
        >
          Heroes App
        </span>
      </a>
      <ul
        className="nav col-12 col-lg-auto me-lg-auto mb-2 justify-content-center mb-md-0"
      >
        <li>
          <a
            aria-current="page"
            className="nav-link px-2 text-secondary"
            href="/home"
            onClick={[Function]}
          >
            Home
          </a>
        </li>
        <li>
          <a
            className="nav-link px-2 text-white"
            href="/marvel"
            onClick={[Function]}
          >
            Marvel
          </a>
        </li>
        <li>
          <a
            className="nav-link px-2 text-white"
            href="/dc"
            onClick={[Function]}
          >
            Dc
          </a>
        </li>
      </ul>
      <div
        className="text-end"
      >
        <span
          className="me-4"
        >
          Name
        </span>
        <button
          className="btn btn-outline-light me-2"
          onClick={[Function]}
          type="button"
        >
          Logout
        </button>
      </div>
    </div>
  </div>
</header>
`;
